Qute template with null default parameter fails with unclear NPE

### Describe the bug

As a developer, it may seem like a good idea to define a Qute template parameter with an explicit null default. This would ensure that the parameter is always available, even when templates are sometimes rendered without this parameter provided.


Example:

```qute
{@String name=null}

Hello {name} from Qute!
```

### Expected behavior

This should work or should provide a clear error message outlining what template may have an error, ideally including the parameter name that failed.

### Actual behavior

The code fails with a generic stacktrace:
```
2025-09-07 20:30:34,651 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.qute.deployment.QuteProcessor#validateExpressions threw an exception: java.lang.NullPointerException: Cannot invoke "org.jboss.jandex.Type.kind()" because "type" is null
        at io.quarkus.qute.deployment.Types.box(Types.java:189)                                                                                                                                  
        at io.quarkus.qute.deployment.Types$AssignabilityCheck.isAssignableFrom(Types.java:151)                                                                                                  
        at io.quarkus.qute.deployment.QuteProcessor.validateDefaultValuesOfParameterDeclarations(QuteProcessor.java:1142)                                                                        
        at io.quarkus.qute.deployment.QuteProcessor.validateExpressions(QuteProcessor.java:1081)                                                                                                 
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)                                                                                                    
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:874)                                                                                                             
        at io.quarkus.builder.BuildContext.run(BuildContext.java:255)                                                                                                                            
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)                                                                                                                    
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)                                                                                               
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)                                                                                                     
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)                                                                                                
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)                                                                                               
        at java.base/java.lang.Thread.run(Thread.java:840)                                                                                                                                       
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)                                                                                                                               
                                                                                                                                                                                                 
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:376)                                                                                                  
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:293)                                                                             
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:61)                                                                              
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:89)                                                                                                 
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:432)                                                                                                    
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:55)                                                                                                     
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:143)                                                                                                      
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:98)                                                                                       
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:104)                                                                                                                     
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:67)                                                                                                                       
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors                                                                                                          
        [error]: Build step io.quarkus.qute.deployment.QuteProcessor#validateExpressions threw an exception: java.lang.NullPointerException: Cannot invoke "org.jboss.jandex.Type.kind()" because "type" is null                                                                                                                                                                                  
        at io.quarkus.qute.deployment.Types.box(Types.java:189)                                                                                                                                  
        at io.quarkus.qute.deployment.Types$AssignabilityCheck.isAssignableFrom(Types.java:151)                                                                                                  
        at io.quarkus.qute.deployment.QuteProcessor.validateDefaultValuesOfParameterDeclarations(QuteProcessor.java:1142)                                                                        
        at io.quarkus.qute.deployment.QuteProcessor.validateExpressions(QuteProcessor.java:1081)                                                                                                 
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)                                                                                                    
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:874)                                                                                                             
        at io.quarkus.builder.BuildContext.run(BuildContext.java:255)                                                                                                                            
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)                                                                                                                    
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)                                                                                               
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)                                                                                                     
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)                                                                                                
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)                                                                                               
        at java.base/java.lang.Thread.run(Thread.java:840)                                                                                                                                       
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)                                                                                                                               
                                                                                                                                                                                                 
        at io.quarkus.builder.Execution.run(Execution.java:122)                                                                                                                                  
        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:78)                                                                                                       
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)                                                                                                                 
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:372)                                                                                                  
        ... 9 more                                                                                                                                                                               
Caused by: java.lang.NullPointerException: Cannot invoke "org.jboss.jandex.Type.kind()" because "type" is null                                                                                   
        at io.quarkus.qute.deployment.Types.box(Types.java:189)                                                                                                                                  
        at io.quarkus.qute.deployment.Types$AssignabilityCheck.isAssignableFrom(Types.java:151)                                                                                                  
        at io.quarkus.qute.deployment.QuteProcessor.validateDefaultValuesOfParameterDeclarations(QuteProcessor.java:1142)                                                                        
        at io.quarkus.qute.deployment.QuteProcessor.validateExpressions(QuteProcessor.java:1081)                                                                                                 
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)                                                                                                    
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:874)                                                                                                             
        at io.quarkus.builder.BuildContext.run(BuildContext.java:255)                                                                                                                            
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)                                                                                                                    
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)                                                                                               
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)                                                                                                     
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)                                                                                                
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)                                                                                               
        at java.base/java.lang.Thread.run(Thread.java:840)                                                                                                                                       
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

This does not indicate where the root cause of this error may be, making it difficult to resolve the issue.

### How to Reproduce?

Reproducer https://github.com/yholkamp/quarkus-null-repro or this file:

[quarkus-null-repro-main.zip](https://github.com/user-attachments/files/22198098/quarkus-null-repro-main.zip)

Run `quarkus dev`

### Output of `uname -a` or `ver`

Linux ydesk 6.14.0-29-generic #29~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Aug 14 16:52:50 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.15" 2025-04-15 OpenJDK Runtime Environment Temurin-17.0.15+6 (build 17.0.15+6) OpenJDK 64-Bit Server VM Temurin-17.0.15+6 (build 17.0.15+6, mixed mode, sharing)

### Quarkus version or git rev

3.26.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9

### Additional information

_No response_

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
@@ -1113,6 +1113,9 @@ public class QuteProcessor {
             if (defaultValue != null) {
                 MatchResult match;
                 if (defaultValue.isLiteral()) {
+                    if (defaultValue.getLiteral() == null)
+                        // Skip validation for `null`
+                        continue;
                     match = new MatchResult(assignabilityCheck);
                     setMatchValues(match, defaultValue, generatedIdsToMatches, index);
                 } else {

--- a/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/typesafe/ParamDeclarationDefaultValueTest.java
+++ b/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/typesafe/ParamDeclarationDefaultValueTest.java
@@ -19,7 +19,12 @@ public class ParamDeclarationDefaultValueTest {
             .withApplicationRoot((jar) -> jar
                     .addClass(MyEnum.class)
                     .addAsResource(new StringAsset(
-                            "{@io.quarkus.qute.deployment.typesafe.ParamDeclarationDefaultValueTest$MyEnum myEnum=MyEnum:BAR}{myEnum}"),
+                            """
+                                    {@io.quarkus.qute.deployment.typesafe.ParamDeclarationDefaultValueTest$MyEnum myEnum=MyEnum:BAR}
+                                    {@String name=null}
+                                    {myEnum}
+                                    {#if name != null}NOK!{/}
+                                    """),
                             "templates/myEnum.html"));
 
     @Inject
@@ -27,7 +32,7 @@ public class ParamDeclarationDefaultValueTest {
 
     @Test
     public void testDefaultValue() {
-        assertEquals("BAR", myEnum.render());
+        assertEquals("BAR", myEnum.render().strip());
     }
 
     @TemplateEnum

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/ExpressionImpl.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/ExpressionImpl.java
@@ -37,10 +37,11 @@ final class ExpressionImpl implements Expression {
         if (literal == null) {
             throw new IllegalArgumentException("Literal must not be null");
         }
-        return new ExpressionImpl(id, null,
-                Collections.singletonList(new PartImpl(literal,
-                        value != null ? Expressions.typeInfoFrom(value.getClass().getName()) : null)),
-                value, origin);
+        String typeInfo = null;
+        if (value != null) {
+            typeInfo = Expressions.typeInfoFrom(value.getClass().getName());
+        }
+        return new ExpressionImpl(id, null, List.of(new PartImpl(literal, typeInfo)), value, origin);
     }
 
     static Integer syntheticId() {

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/SectionHelperFactory.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/SectionHelperFactory.java
@@ -162,6 +162,10 @@ public interface SectionHelperFactory<T extends SectionHelper> {
 
     interface BlockInfo extends ParserDelegate, WithOrigin {
 
+        default boolean isMainBlock() {
+            return MAIN_BLOCK_NAME.equals(getLabel());
+        }
+
         String getLabel();
 
         /**

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/SetSectionHelper.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/SetSectionHelper.java
@@ -139,7 +139,7 @@ public class SetSectionHelper implements SectionHelper {
 
         @Override
         public Scope initializeBlock(Scope previousScope, BlockInfo block) {
-            if (block.getLabel().equals(MAIN_BLOCK_NAME)) {
+            if (block.isMainBlock()) {
                 Scope newScope = new Scope(previousScope);
                 for (Entry<String, String> e : block.getParameters().entrySet()) {
                     String key = e.getKey();

--- a/independent-projects/qute/core/src/test/java/io/quarkus/qute/ParamDeclarationDefaultValueTest.java
+++ b/independent-projects/qute/core/src/test/java/io/quarkus/qute/ParamDeclarationDefaultValueTest.java
@@ -19,6 +19,7 @@ public class ParamDeclarationDefaultValueTest {
         assertDefaultValue(engine.parse("{@java.lang.String val= 'foo'}\n{val.toUpperCase}"), "FOO");
         assertDefaultValue(engine.parse("{@java.lang.String val='foo and bar'}\n{val.toUpperCase}"), "FOO AND BAR");
         assertDefaultValue(engine.parse("{@java.lang.String val= 'foo and bar'}\n{val.toUpperCase}"), "FOO AND BAR");
+        assertDefaultValue(engine.parse("{@java.lang.String val=null}\n{val == null ? 'ok' : 'nok'}"), "ok");
     }
 
     @Test
@@ -56,12 +57,14 @@ public class ParamDeclarationDefaultValueTest {
         assertEquals(expectedOutput, template.render());
         Expression fooExpr = template.getExpressions().stream().filter(e -> e.isLiteral()).findFirst().orElse(null);
         assertNotNull(fooExpr);
-        assertEquals("|java.lang.String|", fooExpr.collectTypeInfo());
-        Expression valExpr = template.getExpressions().stream().filter(e -> e.toOriginalString().equals("val.toUpperCase"))
-                .findAny().orElse(null);
-        assertNotNull(valExpr);
-        assertEquals("|java.lang.String|.toUpperCase", valExpr.collectTypeInfo());
-        assertEquals(2, valExpr.getOrigin().getLine());
+        if (fooExpr.getLiteral() != null) {
+            assertEquals("|java.lang.String|", fooExpr.collectTypeInfo());
+            Expression valExpr = template.getExpressions().stream().filter(e -> e.toOriginalString().equals("val.toUpperCase"))
+                    .findAny().orElse(null);
+            assertNotNull(valExpr);
+            assertEquals("|java.lang.String|.toUpperCase", valExpr.collectTypeInfo());
+            assertEquals(2, valExpr.getOrigin().getLine());
+        }
     }
 
 }

